agent/cache-types/service_checks_test.go

```diff
@@ -0,0 +1,178 @@
+		t.Fatalf("failed to add check: %v", err)
+	}
+	result2, err := typ.Fetch(
+		cache.FetchOptions{LastResult: &result},
+		&cachetype.ServiceHTTPChecksRequest{ServiceID: svcState.Service.ID},
+	)
+	if err != nil {
+		t.Fatalf("failed to fetch: %v", err)
+	}
+	if result2.Index != 2 {
+		t.Fatalf("expected index of 2 after second request, got %d", result2.Index)
+	}
+
+	got, ok = result2.Value.(*[]structs.CheckType)
+	if !ok {
+		t.Fatalf("fetched value of wrong type, got %T, want *[]structs.CheckType", got)
+	}
+	want = chkTypes[0:2]
+	for i, c := range *got {
+		require.Equal(t, c, *want[i])
+	}
+
+	// Removing GRPC check should yield HTTP check in Fetch
+	if err := a.RemoveCheck(chkTypes[1].CheckID); err != nil {
+		t.Fatalf("failed to add check: %v", err)
+	}
+	result3, err := typ.Fetch(
+		cache.FetchOptions{LastResult: &result2},
+		&cachetype.ServiceHTTPChecksRequest{ServiceID: svcState.Service.ID},
+	)
+	if err != nil {
+		t.Fatalf("failed to fetch: %v", err)
+	}
+	if result3.Index != 3 {
+		t.Fatalf("expected index of 3 after third request, got %d", result3.Index)
+	}
+
+	got, ok = result3.Value.(*[]structs.CheckType)
+	if !ok {
+		t.Fatalf("fetched value of wrong type, got %T, want *[]structs.CheckType", got)
+	}
+	want = chkTypes[0:1]
+	for i, c := range *got {
+		require.Equal(t, c, *want[i])
+	}
+
+	// Fetching again should yield no change in result nor index
+	result4, err := typ.Fetch(
+		cache.FetchOptions{LastResult: &result3},
+		&cachetype.ServiceHTTPChecksRequest{ServiceID: svcState.Service.ID, MaxQueryTime: 100 * time.Millisecond},
+	)
+	if err != nil {
+		t.Fatalf("failed to fetch: %v", err)
+	}
+	if result4.Index != 3 {
+		t.Fatalf("expected index of 3 after fetch timeout, got %d", result4.Index)
+	}
+
+	got, ok = result4.Value.(*[]structs.CheckType)
+	if !ok {
+		t.Fatalf("fetched value of wrong type, got %T, want *[]structs.CheckType", got)
+	}
+	want = chkTypes[0:1]
+	for i, c := range *got {
+		require.Equal(t, c, *want[i])
+	}
+}
+
+type mockAgent struct {
+	state   *local.State
+	pauseCh <-chan struct{}
+	checks  []structs.CheckType
+}
+
+func newMockAgent() *mockAgent {
+	m := mockAgent{
+		state:   local.NewState(local.Config{NodeID: "host"}, nil, new(token.Store)),
+		pauseCh: make(chan struct{}),
+		checks:  make([]structs.CheckType, 0),
+	}
+	m.state.TriggerSyncChanges = func() {}
+	return &m
+}
+
+func (m *mockAgent) ServiceHTTPChecks(id string) []structs.CheckType {
+	return m.checks
+}
+
+func (m *mockAgent) LocalState() *local.State {
+	return m.state
+}
+
+func (m *mockAgent) SyncPausedCh() <-chan struct{} {
+	return m.pauseCh
+}
+
+func (m *mockAgent) AddCheck(check structs.CheckType) error {
+	m.checks = append(m.checks, check)
+	return nil
+}
+
+func (m *mockAgent) RemoveCheck(id types.CheckID) error {
+	new := make([]structs.CheckType, 0)
+	for _, c := range m.checks {
+		if c.CheckID != id {
+			new = append(new, c)
+		}
+	}
+	m.checks = new
+	return nil
+}
```

Comment:
Other cache type test have a check that asserts the correct response type handling which was a real bug at one point so it's probably worth duplicating as a sanity check here. e.g. https://github.com/hashicorp/consul/blob/74461406e01ee7c73387bcf4b91bedad1802dadf/agent/cache-types/config_entry_test.go#L56